src/block/Composter.php

```diff
@@ -0,0 +1,203 @@
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\block\utils\CompostFactory;
+use pocketmine\data\runtime\RuntimeDataDescriber;
+use pocketmine\event\block\ComposterEmptyEvent;
+use pocketmine\event\block\ComposterFillEvent;
+use pocketmine\event\block\ComposterMatureEvent;
+use pocketmine\item\Item;
+use pocketmine\item\VanillaItems;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use pocketmine\world\particle\CropGrowthEmitterParticle;
+use pocketmine\world\sound\ComposterEmptySound;
+use pocketmine\world\sound\ComposterFillSound;
+use pocketmine\world\sound\ComposterFillSuccessSound;
+use pocketmine\world\sound\ComposterReadySound;
+use function max;
+use function mt_rand;
+
+class Composter extends Transparent{
+
+	public const EMPTY_LEVEL = 0;
+	public const FULL_LEVEL = 7;
+	public const COLLECTING_LEVEL = 8;
+
+	protected int $fillLevel = self::EMPTY_LEVEL;
+
+	protected function describeBlockOnlyState(RuntimeDataDescriber $w) : void{
+		$w->boundedInt(4, self::EMPTY_LEVEL, self::COLLECTING_LEVEL, $this->fillLevel);
+	}
+
+	protected function recalculateCollisionBoxes() : array{
+		$empty_layer = (max(1, 15 - 2 * $this->fillLevel) - (int) ($this->fillLevel === 0)) / 16;
+		$boxes = [AxisAlignedBB::one()->trim(Facing::UP, $empty_layer)];
+
+		foreach(Facing::HORIZONTAL as $side){
+			$boxes[] = AxisAlignedBB::one()->trim(Facing::opposite($side), 14 / 16);
+		}
+		return $boxes;
+	}
+
+	public function isEmpty() : bool{
+		return $this->fillLevel === self::EMPTY_LEVEL;
+	}
+
+	public function isFull() : bool{
+		return $this->fillLevel === self::FULL_LEVEL;
+	}
+
+	public function canCollect() : bool{
+		return $this->fillLevel === self::COLLECTING_LEVEL;
+	}
+
+	public function getFillLevel() : int{
+		return $this->fillLevel;
+	}
+
+	/** @return $this */
+	public function setFillLevel(int $fillLevel) : self{
+		if($fillLevel < 0 || $fillLevel > self::COLLECTING_LEVEL){
+			throw new \InvalidArgumentException("Layers must be in range " . self::EMPTY_LEVEL . " ... " . self::COLLECTING_LEVEL);
+		}
+		$this->fillLevel = $fillLevel;
+		return $this;
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if($this->isFull()){
```

Comment:
How can it be full when it's about to be placed?

Reply:
> How can it be full when it's about to be placed?

In vanilla, if you place a composter with a level of 7 (Full)... It then got scheduled to make collectable composter

Reply:
I was originally wondering how to place a block that already had a level filled, but then I remembered the click wheel.

Reply:
> I was originally wondering how to place a block that already had a level filled, but then I remembered the click wheel.

might I ask what is this "click wheel"?

Reply:
I called it that, but it's the mouse wheel, when you click on it.